Allow MACAddressPolicy=persistent for all virtual devices

keszybz opened the pull request:
This is another approach to #3374. See the last patch for an extensive description.

keszybz commented:
I restarted bionic-{amd64,i386} after merging #11381.

keszybz commented:
It passed now, fwiw.

yuwata commented:
LGTM, and works fine. But, let's wait for other reviewers.

keszybz commented:
I'll set the v242 milestone for it — the PR should either be accepted early in the v242 cycle or rejected.

poettering commented on src/libsystemd-network/network-internal.c:
```diff
@@ -40,15 +40,18 @@ const char *net_get_name(sd_device *device) {
         name = net_get_name(device);
         if (!name)
-                return -ENOENT;
+                if (sd_device_get_property_value(device, "ID_BUS", NULL) == -ENOENT)
```
I am not sure i understand the rationale for this check. I mean, if a device has zero identifying info, but is a hw device, why not just use the "eth0" name? I mean, giving eth0 (whatever it might be) the same MAC address sounds like an OK thing to me, no?

I mean, I like simplicity: i'd really just use the sysname here if no proper identification is available, and doesn't sound so bad to me, not sure why adding the extra complexity of the if check here...

poettering commented on src/libsystemd-network/network-internal.c:
```diff
@@ -40,15 +40,18 @@ const char *net_get_name(sd_device *device) {
+                if (sd_device_get_property_value(device, "ID_BUS", NULL) == -ENOENT)
+                        (void) sd_device_get_sysname(device, &name);
+        if (!name)
+                return -ENODATA;
```
Can we add some explanation here in a comment what this is doing, and why? Also, I think it would be a great idea if we'd have debug logging here about what we are doing here, since I am pretty sure it's highly interesting to figure out sometimes how a specific MAC came to be...

poettering commented:
I like the general approach, but I'd simplify it, see above. I guess this also needs a rebase once #11449 has been merged.

tomty89 commented:
> net_get_name() is called from dhcp_ident_set_iaid() so I didn't change net_get_name() like in grafi-tt's patch, but net_get_unique_predictable_data().

That's why I don't like it. Doesn't that indicate it is "trying too hard"?

keszybz commented:
> That's why I don't like it. Doesn't that indicate it is "trying too hard"?

It's about preserving stability. In this particular case, it's just the stability of a network identifier, used to generate DHCP addresses, so if it were to change, it wouldn't be such a big issue. But it was simple enough to keep it unchanged. If we think the code should be simplified, then this can be done as a separate PR. But then I'd make this code path reuse `net_get_unique_predictable_data` and fully drop the custom seed creation.

tomty89 commented:
To me it isn't just about simplicity, but definition.

> The only case which is not covered is devices connected to a bus (i.e. physical
devices), that don't have any identifying information. While setting a fixed
mac address to "bond1" seems OK, doing the same each time a USB dongle is
plugged in and gets the name eth0 seems wrong. So physical devices are
excluded. We can always extend the policy to cover them later.

To be honest, I am not seeing any logic in this. So what exactly do we consider safe/right to be used as the seed of a persistent MAC now?

It appears to me what you are trying to do here is, "we don't know if the name is suitable for seeding, but since only virtual devices have problem with the *universal* MACAddressPolicy=persistent we ship, so whatever, let's just whitelist them".

> net_get_name() is called from dhcp_ident_set_iaid() so I didn't change
net_get_name() like in grafi-tt's patch, but net_get_unique_predictable_data().
>
> net_get_unique_predictable_data() is called from get_mac() in link-config.c
and sd_ipv4ll_set_address_seed(), so both of those code paths are affected
and will now get data in some cases where they errored out previously.

Similarly, "we don't know if the new thing we do is right, let's just affect as less thing as we can".

To preserve stability, you can just do this, no?
```
const char *net_get_name(sd_device *device) {
        const char *name, *field;

        assert(device);

        /* fetch some persistent data unique (on this machine) to this device */
        FOREACH_STRING(field, "ID_NET_NAME_ONBOARD", "ID_NET_NAME_SLOT", "ID_NET_NAME_PATH", "ID_NET_NAME_MAC")
                if (sd_device_get_property_value(device, field, &name) >= 0)
                        return name;

        if (sd_device_get_sysname(device, &name) >= 0)
                return name;

        return NULL;
}
```
The question is just whether it's fine / the right thing to do.

mrc0mmand commented:
Two udev tests are now keep failing with this PR:

ninja-test: https://ci.centos.org/job/systemd-pr-build/4444/artifact/systemd-centos-ci/artifacts_2rQPzp/testsuite-logs.AHt/ninja-test_FAIL.log
TEST-24-UNIT-TESTS: https://ci.centos.org/job/systemd-pr-build/4444/artifact/systemd-centos-ci/artifacts_2rQPzp/testsuite-logs.AHt/TEST-24-UNIT-TESTS_FAIL.log
Index: https://ci.centos.org/job/systemd-pr-build/4444/artifact/systemd-centos-ci/index.html

keszybz commented:
I pushed another commit to make test-libudev spout more output.

poettering commented:
lgtm

poettering commented:
hmm, the travis CI fails...

poettering commented:
hmm, I am a bit torn, i do see @tomty89's point that it is a bit weird to override the MAC address of API-created interfaces like we do right now. But also: adding stable MACs derived from /etc/machine-id is only something we can reasonably do, as the other participants mostly don't acknowledge the concept of /etc/machine-id... but dunno, maybe that's not reason enough?

Hence, dunno, i think this PR is OK, but I can also see @tomty89's point...

I also still think the ID_BUS conditionalization is not really too convincing, I have a slight preferences to drop that conditoinalization. My thinking goes like this: anyone who has more than one iface really should want stable interface names, otherwise his life is constant pain. People who stick to names like "eth0" are hence usually those who have only one interface and thus don't care about the name much. But if that's the case, then having a stable MAC address for this single ethernet port actually *is* a good thing, too, and hence in that deriving the name from eth0+machine-id doesn't sound bad at all, but good.

keszybz commented:
One test failed because of hwdb is not installed, and I added an assert that the test passes. I'll remove this change.

But more importantly, the refactoring was borked. This is now fixed, and I added yet more debugging information, including warnings if we encounter `addr_assign_type` that we don't know.

poettering commented:
lgtm

keszybz commented:
`E: Malformed entry 1 in list file /etc/apt/sources.list.d/autopkgtest.list (URI parse)`

borkmann commented:
@keszybz @tomty89 @pasky This unfortunately breaks network connectivity for us as well when users upgrade to a newer systemd. :-( Having to ask users to disable the default behavior via https://github.com/systemd/systemd/issues/3374#issuecomment-288882355 is a very unfortunate user experience. Why couldn't such persistent MAC address generation be done in the kernel directly for all virtual devices? Right now they generate a random one anyway, and this would actually be transparent to users.

keszybz commented:
> Why couldn't such persistent MAC address generation be done in the kernel directly for all virtual devices e.g. via sysctl switch? Right now they generate a random one anyway, and this would actually be transparent to users.

The MAC address generated by udev hashes both the interface name (or other stable data) and the machine-id. This way different machines with the same configuration do not get the same addresses, yet the address can be stable over reboots. The kernel does not have this kind of information.

@borkmann see #12807. Let's continue the discussion there.

borkmann commented:
> > Why couldn't such persistent MAC address generation be done in the kernel directly for all virtual devices e.g. via sysctl switch? Right now they generate a random one anyway, and this would actually be transparent to users.
> 
> The MAC address generated by udev hashes both the interface name (or other stable data) and the machine-id. This way different machines with the same configuration do not get the same addresses, yet the address can be stable over reboots. The kernel does not have this kind of information.

Right, you need to have a one-time generated, persistent (across reboots) blob of random data for mixing into the mac address generation. This could still be implementable in kernel, takes a bit more complexity, but would be transparent. Anyway, I tried to read up on the design, but couldn't find a description for it yet from the PR comments (could you help me pointing to it, thx). Given this causes regressions all over the board from the comments in the PRs, what is the rationale to have the stable MAC addresses for virtual devices such as veth devs _by default_ and not as opt-in?
